### PR TITLE
Bluetooth: Controller: Maximize BIG event length and preempt PTO & CTRL

### PR DESCRIPTION
**Pull request by cvinayak**
Maximize BIG event length to extend upto ISO interval, and
allow PTO and Control subevents to be pre-emptible.

depends on https://github.com/zephyrproject-rtos/zephyr/pull/59180

### CODE REVIEW

**Comment by carlescufi** [comment]
@kruithofa @Thalley PTAL

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c [hunk 1]
```diff
@@ -275,14 +282,62 @@ uint8_t ll_big_create(uint8_t big_handle, uint8_t adv_handle, uint8_t num_bis,
...
+	}
+
+	/* Calculate overheads due to extended advertising. */
+	if (aux && aux->is_started) {
+		ticks_slot_aux = aux->ull.ticks_slot;
+		if (IS_ENABLED(CONFIG_BT_CTLR_LOW_LAT)) {
+			ticks_slot_overhead = MAX(aux->ull.ticks_active_to_start,
+						  aux->ull.ticks_prepare_to_start);
+		} else {
+			ticks_slot_overhead = 0;
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:300]:
```suggestion
			ticks_slot_overhead = 0U;
```

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c [hunk 2]
```diff
@@ -275,14 +282,62 @@ uint8_t ll_big_create(uint8_t big_handle, uint8_t adv_handle, uint8_t num_bis,
...
+
+	/* Calculate overheads due to periodic advertising. */
+	sync = HDR_LLL2ULL(lll_adv_sync);
+	if (sync->is_started) {
+		ticks_slot_sync = sync->ull.ticks_slot;
+		if (IS_ENABLED(CONFIG_BT_CTLR_LOW_LAT)) {
+			ticks_slot_overhead = MAX(sync->ull.ticks_active_to_start,
+						  sync->ull.ticks_prepare_to_start);
+		} else {
+			ticks_slot_overhead = 0;
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:315]:
```suggestion
			ticks_slot_overhead = 0U;
```

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c [hunk 3]
```diff
@@ -891,6 +946,11 @@ void ull_adv_iso_stream_release(struct ll_adv_iso_set *adv_iso)
 	lll->adv = NULL;
 }
 
+uint32_t ull_adv_iso_max_time_get(struct ll_adv_iso_set *adv_iso)
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:949]:
```suggestion
uint32_t ull_adv_iso_max_time_get(const struct ll_adv_iso_set *adv_iso)
```

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c [hunk 4]
```diff
@@ -944,33 +1004,49 @@ static uint8_t ptc_calc(const struct lll_adv_iso *lll, uint32_t event_spacing,
 	return 0U;
 }
 
+static uint32_t adv_iso_time_get(struct ll_adv_iso_set *adv_iso, bool max)
+{
+	struct lll_adv_iso *lll_iso;
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:1009]:
```suggestion
static uint32_t adv_iso_time_get(const struct ll_adv_iso_set *adv_iso, bool max)
{
	const struct lll_adv_iso *lll_iso;
```

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c [hunk 5]
```diff
@@ -944,33 +1004,49 @@ static uint8_t ptc_calc(const struct lll_adv_iso *lll, uint32_t event_spacing,
...
+
+	if (IS_ENABLED(CONFIG_BT_CTLR_ADV_ISO_RESERVE_MAX) || max) {
+		time_us = (pdu_spacing * lll_iso->nse * lll_iso->num_bis) +
+			  ctrl_spacing;
+	} else {
+		time_us = pdu_spacing * ((lll_iso->nse * lll_iso->num_bis) -
+					 lll_iso->ptc);
+	}
+
+	time_us += EVENT_OVERHEAD_START_US + EVENT_OVERHEAD_END_US;
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:1030]: Any chance that `time_us` can overflow? 

You are multiplying a `uint32_t` by another another number and storing the result in another `uint32_t`.

File: subsys/bluetooth/controller/ll_sw/ull_adv_sync.c [hunk 6]
```diff
@@ -225,7 +225,7 @@ uint8_t ll_adv_sync_param_set(uint8_t handle, uint16_t interval, uint16_t flags)
 }
 
 #if defined(CONFIG_BT_CTLR_ADV_ISO) && defined(CONFIG_BT_TICKER_EXT_EXPIRE_INFO)
-void ull_adv_iso_created(struct ll_adv_sync_set *sync)
+void ull_adv_sync_iso_created(struct ll_adv_sync_set *sync)
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:228]:
```suggestion
void ull_adv_sync_iso_created(const struct ll_adv_sync_set *sync)
```

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_adv_sync.c (hunk 6 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv_sync.c]: done.

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c (hunk 2 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv_iso.c]: done

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c (hunk 1 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv_iso.c]: done

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c (hunk 3 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv_iso.c]: done

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c (hunk 4 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv_iso.c]: done

File: subsys/bluetooth/controller/ll_sw/ull_adv_iso.c (hunk 5 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv_iso.c]: `time_us` will not overflow for the Max PDU size of 251 bytes on S8 coding, with NSE of 31 and 31 BISes. Added comments in the code calculating the required bits for the maximum radio event time.

File: subsys/bluetooth/controller/ll_sw/ull_adv_sync.c (hunk 6 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:225]: Had to remove the `const` as `sync` passed as ticker callback parameter which can modify the `sync` object/memory.

**Review by Thalley — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**